### DISCUSSION TITLE
Spring Integration - JmsOutboundGatway timeouts

### DISCUSSION THREAD

**Discussion by robertgosz** [spring-projects/spring-integration#4005] [Q&A] [answered]
Hi,
I have a question regarding JmsOutboundGateway and timeouts. I was using Spring Boot 2.3.12.RELEASE with SpringIntegration and JmsOutboundGateway to send and receive async messages using IBM MQ. In the gateway configuration i used to set receiveTimeout value and if there was no reply on the ReplyDestination queue within this time JmsTimeoutException was thrown and OnFailure callback method was called. That was expected behaviour, nothing to talk about.
After switching to Spring Boot 2.6.6 I have noticed the timeouts not being thrown until the application was shut down. I did some testing and figured out that if I set replyTimeout in the @Gatway definition then I can make it "work", but the behaviour is quite misleading for me. For example - when I set replyTimeout to 60s and receiveTimeout to 5s the JmsTimeoutException is thrown with a message: "org.springframework.integration.jms.JmsTimeoutException: No reply in 5000 ms", but it's being thrown after 60s, not 5s. If I set both to 60s then after 60s onSuccess is being called with null payload instead of onFailure. Similar thing happens when I set receiveTimeout to 60s and replyTimeout to 5s - after 5s onSuccess is being called with null payload instead of onFailure with timeout exception. Can anyone explain what happens there and if it's the expected behaviour ? I could not find any reasonable explanation in the documentation. Maybe I have missed or misunderstood something. Thanks in advance!

**Comment by artembilan** [comment, marked as answer]
Ho @robertgosz !

Thank you for the report!

Any chances that you can share with us a simple project to play with and reproduce?

**Reply by robertgosz** [reply]
Hi,
thanks for your response. I have created a small test app here: https://github.com/robertgosz/SpringIntegrationTest
I have added my findings inside MessageGateway.java. The receiveTimeout is set in application.properties.
Unfortunately it's IBM MQ based setup as I haven't tried other messaging providers. I hope it will be useful anyway.
Any feedback appreciated!

Best Regards
Robert

**Reply by artembilan** [reply]
Let's see if this one helps you:
```
@MessagingGateway(asyncExecutor = AnnotationConstants.NULL)
```
!

The point is that you now send and receive via `@Gateway` in async manner and at the same time your `JmsOutboundGateway` produces a `Future` for its reply.
I guess in that old Spring Integration version we didn't have a special handling for `ListenableFuture` so it was not processed on an `asyncExecutor`.
The current goal is to return a JMS `Future` reply as is.

Unfortunately your sample project is a bit involved with IBM MQ and Tomcat to make it as simple as possible to play with ActiveMQ, for example.
You also need to see into upgrading to the latest Spring Boot: https://spring.io/projects/spring-boot#learn

**Reply by robertgosz** [reply]
Hi Artem,

Tried with '@MessagingGateway(asyncExecutor = AnnotationConstants.NULL)' and in this case sendAndReceiveMessage does not return immediately. It freezes for replyTimeout ms and then returns null instead of future. Sometimes (don't know why the result is not the same each time):

org.springframework.integration.jms.JmsTimeoutException: No reply in 5000 ms
	at org.springframework.integration.jms.JmsOutboundGateway.expire(JmsOutboundGateway.java:1184)
	at org.springframework.integration.jms.JmsOutboundGateway.lambda$createFuture$12(JmsOutboundGateway.java:1173)

is being thrown (again, the 5000 ms is wrong there). If replyTimeout is not defined - nothing happens - no exception thrown. Either way 'LOG.info("______ Message sent ______");' is never executed. Although the message is being sent. Tested with SpringBoot 2.7.8.

I know this setup is MQ bound, sorry, that's what I have. I'll try to reproduce it with ActiveMQ if possible.

**Reply by robertgosz** [reply]
I have added ActiveMQ version of the test app: https://github.com/robertgosz/SpringIntegrationTestActiveMQ. Should be easy to run with embedded Tomcat and out of the box Classic installation of ActiveMQ (just create queues) and without proprietary messaging provider. Initial testing suggests that the behaviour is exactly the same as in IBM MQ case.

**Reply by artembilan** [reply]
Thank you! Will look into that today or tomorrow.
Would you mind, please, remind me what exactly is an expected behavior?

**Reply by robertgosz** [reply]
Thanks. I really appreciate your help.
In my mind the 'expected behaviour' is to **get timeoutException after specified amount of time if there is no response on the response queue and having onFailure called**. We might call the old SpringBoot 2.3.12.RELEASE behaviour the 'expected one' as that's what I was used to. And once again what I found different in those newer versions of SpringBoot (2.6, 2.7):
1. If the JmsTimeoutException is thrown - the message from the expire method is misleading (exception is really thrown after replyTimeout ms and the message suggests that it is thrown after receiveTimeout ms). 
2. When receiveTimeout is higher or equal to replyTimeout - no exception is thrown - instead onSuccess is called with null payload after replyTimeout ms.
3. Even when replyTimeout is higher than receiveTimeout the behaviour is not always the same. Sometimes exception is thrown, sometimes onSuccess is called. That's kind of strange IMHO. 
4. If no replyTimeout is provided in the Gateway annotation - timeouts are never thrown (receiveTimeout alone is being ignored).

**Reply by artembilan** [reply]
Ok. Looking into that. 
Have just simplified your application to deal with an embedded ActiveMQ.
I also don't see a reason in the Tomcat at all...

**Reply by robertgosz** [reply]
That's great, 
I have just cut the examples out of a bigger application that I used to run with Tomcat, no any other reason for it to be there. 
You can push your changes back if you want. And to be clear - I'm not sure if it's an issue, or maybe I should have configured something differently while migrating to higher version of SpringBoot. That's why I started discussion instead of creating an issue ticket.

Best Regards
Robert

**Reply by artembilan** [reply]
No, that's OK: it is just hard to concentrate on the problem when we have some extra disturbance.
This is probably OK for you since it is your project, but I have to deal with many similar issues all the time and it is a bit frustrating to get lost in whatever is not relevant.

I can confirm with old Spring Boot version what you see is really an expected behavior.
Here is what is going on:
1. When we call a `@Gateway`, it returns to us a `ListenableFuture` immediately and schedule an async task for it: `return ((AsyncListenableTaskExecutor) this.asyncExecutor).submitListenable(invoker::get);`
2. That `invoker.get()` sends a message downstream and waits for reply.
3. The `MessagingTemplate.sendAndReceive()` populates a `TemporaryReplyChannel` into a `replyChannel` and `errorChannel` message headers.
4. The async `JmsOutboundGateway` performs an `expire()` on the `SettableListenableFuture` since we don't send a reply.
5. At this point we do `future.setException(new JmsTimeoutException("No reply in " + this.receiveTimeout + " ms"));`
6. Since it is an async request-reply, this `JmsOutboundGateway` returns the mentioned `ListenableFuture` as a reply back to its super class.
7. This `AbstractMessageProducingHandler` super class adds callback to the mentioned reply `ListenableFuture` to perform a send to the `replyChannel` or `errorChannel` from headers, respectively.
8. In case of `onFailure()` on this reply `ListenableFuture`(which is essentially our `JmsTimeoutException`) the `MessagingGatewaySupport` extracts this error from the `ErrorMessage` and re-throws it back to the caller.
9. The caller in our case the mentioned before  `invoker.get()` which throws this exception to its caller.
10. And this one is the mentioned scheduled `submitListenable()` which fulfills the top-level, end-user  `ListenableFuture` for your handling in that `TestService`.

That's the report mostly for me since you claim that behavior is different with a latest Spring Integration version, so I'll go baby steps to figure out a difference and fix thereafter.

**Reply by artembilan** [reply]
I cannot push the fix into your repo:
```
remote: Permission to robertgosz/SpringIntegrationTestActiveMQ.git denied to artembilan.
fatal: unable to access 'https://github.com/robertgosz/SpringIntegrationTestActiveMQ.git/': The requested URL returned error: 403
```
But it is very simple.
Add to your `application.properties` this:
```
spring.task.scheduling.pool.size=10
```
This way you don't need to change anything else and an old behavior will be restored.
The problem is that default thread pool is just `1` for shared auto-configured `TaskScheduler` and looks like there is some race condition when that single thread is too busy to let other  tasks in the queue to be performed.
It worked before because Spring Integration used its own `TaskScheduler`, but now it relies on the auto-configured one.
😄

**Reply by robertgosz** [reply]
OMG 🤦‍♂️ 
Such a small but important property I have missed. You are The Man Artem, thanks 🙇‍♂️.
I'm happy there is no issue in the code, and on the other hand - I feel sorry to have used your precious time.
Once again - thank you very much for your support. Lesson learned. 

Best Regards
Robert

**Reply by artembilan** [reply]
No, that's fully OK. 
It is essentially our fault to make such a change in favor of "sensitive default value" which in the end may cause dead locks as we see.
I'll look a bit deeper into code to understand what may take thread from a scheduler just for itself and so on.

I also will chat with Spring Boot team to see if we can lift that `1` thread side to some better value: you are not a first who is facing the problem similar to this.

**Reply by artembilan** [reply]
Well, another way to fix your problem: remove `responseChannel` altogether.
Since you just reply from JMS outbound gateway to the `@Gateway` in the beginning there is no need in extra channel step.
You have it configured as a `QueueChannel` which essentially is blocked by the poller waiting for reply.
And that poller is based on a `TaskScheduler` as well.
That's where you got a dead lock: the thread from `TaskScheduler` is busy waiting for reply.
And we cannot send that reply since, well, thread is busy.
Dead lock 😄

**Reply by robertgosz** [reply]
I did some debugging before posting here and I saw the expiry job being successfully scheduled inside createFuture, but I was wondering why the scheduler never executed the task and did not call expire method. I didn't connect the timeout misbehaviour with scheduler itself causing the issue. Now, I think if it's single and synchronous, then it seems as quite a possible behaviour i guess.

**Reply by robertgosz** [reply]
> Well, another way to fix your problem: remove `responseChannel` altogether. Since you just reply from JMS outbound gateway to the `@Gateway` in the beginning there is no need in extra channel step. You have it configured as a `QueueChannel` which essentially is blocked by the poller waiting for reply. And that poller is based on a `TaskScheduler` as well. That's where you got a dead lock: the thread from `TaskScheduler` is busy waiting for reply. And we cannot send that reply since, well, thread is busy. Dead lock 😄

Hahaha, that's exactly a thought that came to my mind a few minutes ago - we have scheduled expiry,  we are waiting for a reply, we do have some other stuff scheduled, but then - isn't the reply job blocking ? 😄 Cause we're running on one thread 😄

**Reply by robertgosz** [reply]
> Well, another way to fix your problem: remove `responseChannel` altogether. Since you just reply from JMS outbound gateway to the `@Gateway` in the beginning there is no need in extra channel step. 

I use the gateway to talk to another component on the other side of those two queues and for each request I do expect a reply (or timeout). How could I live without responseChannel ?

**Reply by artembilan** [reply]
Right. There is a `TemporaryReplyChannel` in the headers anyway initiated by the `@Gateway`. So, your `JmsOutboundGateway` must not have a `replyChannel` config and the framework will correlate the reply properly with the mentioned `replyChannel` from headers.

**Reply by robertgosz** [reply]
And indeed it worked. 
Less code, same functionality it seems. 
New day, new lesson.
Thanks for your suggestion!
